File: #0064.minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {

        int rowSize = grid.size();
        if (rowSize == 0) {
            return 0;
        }

        int colSize = grid[0].size();
        int x;
        int y;
        for (int i = 0; i < rowSize; i++) {
            for (int j = 0; j < colSize; j++) {
                x = (i - 1 >= 0) ? grid[i - 1][j] : INT_MAX;
                y = (j - 1 >= 0) ? grid[i][j - 1] : INT_MAX;
                grid[i][j] += (std::min(x, y) == INT_MAX) ? 0 : std::min(x, y);
            }
        }
        
        return grid[rowSize - 1][colSize - 1];
    }
};

File: #0064.minimum-path-sum.rs
impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {

        let rowlen = grid.len();
        let collen = grid[0].len();
        let mut board = grid.clone();
        for i in 0..rowlen {
            for j in 0..collen {
                board[i][j] += match (i, j) {
                    (0, 0) => 0,
                    (0, _) => board[i][j - 1],
                    (_, 0) => board[i - 1][j],
                    _ => board[i][j - 1].min(board[i - 1][j]),
                };
            }
        }

        return board[rowlen - 1][collen - 1];
    }
}
